core/src/impl/Kokkos_Profiling.hpp

```diff
@@ -129,6 +130,149 @@ void resume_tools();
 EventSet get_callbacks();
 void set_callbacks(EventSet new_events);
 }  // namespace Experimental
+
+namespace Experimental {
+
+// forward declarations
+size_t get_new_context_id();
+size_t get_current_context_id();
+
+namespace Impl {
+
+static std::map<std::string, Kokkos::Tools::Experimental::TeamSizeTuner>
+    team_tuners;
+
+template <class ExecPolicy, class Functor, typename TagType>
+void tune_policy(const size_t tuning_context, const std::string& label,
+                 ExecPolicy& policy, const Functor& functor,
+                 const TagType& tag){};
+
+template <class Functor, class TagType, class... Properties>
+void tune_policy(const size_t tuning_context, const std::string& label,
+                 Kokkos::TeamPolicy<Properties...>& policy,
+                 const Functor& functor, const TagType& tag) {
+  if (policy.auto_team_size() || policy.auto_vector_length()) {
+    if (team_tuners.find(label) == team_tuners.end()) {
+      team_tuners.emplace(label, Kokkos::Tools::Experimental::TeamSizeTuner(
```

Comment:
Is `label` the unique identifier for a parallel_for instead of `context_id`? See other comment. So all parallel_for with the same label are seen by the tuner as the same?

My worry in that scenario is how do we know we haven't changed any of the task parameters.  We might have a totally new input matrix, e.g., but we wouldn't know that.

Reply:
Okay, this is probably the part of my interface that I like. All kernels with the same label, _in the absence of anything outside of the kernel_, are the same.

So if you write

```c++
for(auto matrix: my_set_of_wildly_different_matrices) {
parallel_for("eat_matrix", AutoTeamPolicyGuy, KOKKOS_LAMBDA (const int i){
  /** stuff */
}
}
```

This is basically useless.

But what you can do is something like

```c++
for(auto matrix: my_set_of_wildly_different_matrices) {
// set Tuning input types to describe the matrix
parallel_for("eat_matrix", AutoTeamPolicyGuy, KOKKOS_LAMBDA (const int i){
  /** stuff */
}
// end the context in which you set those types
}
```

And the tool doing the Tuning actually gets all of that data about your matrix description. Also, if you're lazy, you can just change the label of your kernel based on the matrix, but that would make me sad because the way features compose in this is _so freaking cool_.

Also, I could try to trick some poor soul who works on compilers into trying to figure out what features of a kernel are interesting, and maybe even talk them into generating code to describe their kernels. It could happen!